proxy/iptables: don't call iptables-restore when we don't need to

dcbw opened the pull request:
Diff old and new rules and don't bother calling iptables-restore when nothing has actually changed.

iptables-restore is a very heavy operation and depending on the kernel,
the CPU, and the number of rules to restore, can take a very long time
(~500ms or more).

i7-5600U @ 2.6GHz: 700ms for 1000 services (2+4 cores+threads, kernel 4.6.6)
i7-4790  @ 3.6GHz: 270ms for 1000 services (4+8 cores+threads, kernel 4.6.7)

Other parts of kubernetes (eg kubenet) might be running iptables-restore
too, leading to some pileup as each iptables-restore waits for others to
complete.

Related: https://github.com/kubernetes/kubernetes/issues/26637
Related: https://github.com/kubernetes/kubernetes/issues/33693
Related: https://github.com/kubernetes/kubernetes/pull/35334

<!-- Reviewable:start -->

---
This change is [<img src="https://reviewable.kubernetes.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.kubernetes.io/reviews/kubernetes/kubernetes/36006)
<!-- Reviewable:end -->

@thockin @timothysc @danwinship @eparis

dcbw committed:
util/iptables: add simple iptables rule parser and testcases

We'll need this in the iptables proxy code to figure out when its
pointless to call iptables-restore since things haven't changed.

dcbw committed:
proxy/iptables: don't call iptables-restore when nothing has changed

iptables-restore is a very heavy operation and depending on the kernel,
the CPU, and the number of rules to restore, can take a very long time
(~500ms or more).

i7-5600U @ 2.6GHz: 700ms for 1000 services (2+4 cores+threads, kernel 4.6.6)
i7-4790  @ 3.6GHz: 270ms for 1000 services (4+8 cores+threads, kernel 4.6.7)

Other parts of kubernetes (eg kubenet) might be running iptables-restore
too, leading to some pileup as each iptables-restore waits for others to
complete.

Related: https://github.com/kubernetes/kubernetes/issues/26637
Related: https://github.com/kubernetes/kubernetes/issues/33693
Related: https://github.com/kubernetes/kubernetes/pull/35334

dcbw committed:
proxy/iptables: fix 31 bit masquerade marks

Go assumes int types for numerics that aren't explicitly typed,
leading to:

"constant 2147483648 overflows int"

dcbw commented:
@k8s-bot unit test this https://github.com/kubernetes/kubernetes/pull/36003

dcbw commented:
Rebase to get revert that fixes https://github.com/kubernetes/kubernetes/issues/35898

danwinship commented on pkg/util/iptables/iptables.go:
```diff
@@ -575,3 +576,235 @@ func IsNotFoundError(err error) bool {
+	// find beginning of table
+	for readIndex < len(bytes) {
+		line, n := ReadLine(readIndex, bytes)
+		readIndex = n
```
shouldn't need `n` (likewise below)

dcbw replied:
Fixed (originally copied from somewhere else).

danwinship commented on pkg/util/iptables/iptables.go:
```diff
@@ -575,3 +576,235 @@ func IsNotFoundError(err error) bool {
+		} else if strings.HasPrefix(line, "-A ") {
+			rule, chain, err := parseRule(line, filterChains, filterChainPrefixes)
+			if err != nil {
+				return make(map[Chain][]Rule), err
```
`return nil, err`? (likewise elsewhere)

dcbw replied:
Fixed for the error cases.

danwinship commented on pkg/proxy/iptables/proxier.go:
```diff
@@ -235,7 +236,7 @@ func NewProxier(ipt utiliptables.Interface, sysctl utilsysctl.Interface, exec ut
 	}
 	masqueradeValue := 1 << uint(masqueradeBit)
-	masqueradeMark := fmt.Sprintf("%#08x/%#08x", masqueradeValue, masqueradeValue)
+	masqueradeMark := fmt.Sprintf("%#04x/%#04x", masqueradeValue, masqueradeValue)
```
possibly in the wrong commit?

thockin replied:
This is not right.  At least on my machines, it is actually best-fit:

```
-A FOO -j MARK --set-xmark 0x1/0x1
-A FOO -j MARK --set-xmark 0x10/0x10
-A FOO -j MARK --set-xmark 0x100/0x100
-A FOO -j MARK --set-xmark 0x1000/0x1000
-A FOO -j MARK --set-xmark 0x10000/0x10000
-A FOO -j MARK --set-xmark 0x100000/0x100000
-A FOO -j MARK --set-xmark 0x1000000/0x1000000
-A FOO -j MARK --set-xmark 0x10000000/0x10000000
```

danwinship commented on pkg/proxy/iptables/proxier.go:
```diff
@@ -761,6 +762,89 @@ func (proxier *Proxier) execConntrackTool(parameters ...string) error {
+	newChains, err := parseKubeProxyTableAddRules(table, newLines)
+	if err != nil {
+		return false, err
+	}
```
There's a difference between oldLines failing to parse and newLines failing to parse: oldLines can fail to parse if the administrator mucks around with iptables behind our back, or if the format of iptables-save changes. But newLines can only fail to parse in the event of a programmer error.

dcbw replied:
True, fixed.

dcbw commented:
@k8s-bot node e2e test this issue https://github.com/kubernetes/kubernetes/issues/35983

danwinship commented on pkg/proxy/iptables/proxier.go:
```diff
@@ -1276,13 +1368,24 @@ func (proxier *Proxier) syncProxyRules() {
-		// Revert new local ports.
-		revertPorts(replacementPortsMap, proxier.portsMap)
-		return
+		glog.Warningf("failed to parse iptables filter rules: %v", err)
```
In some failure cases (eg, iptables-save output format changes), every single call to sameIptablesRuleset() will fail... we might want to not log every time?

dcbw replied:
It'll now only warn if parsing the new rules failed, which should be entirely under our control.

danwinship commented on pkg/proxy/iptables/proxier_test.go:
```diff
@@ -257,12 +258,13 @@ func TestExecConntrackTool(t *testing.T) {
+		protocol:            protocol,
+		port:                port,
 		onlyNodeLocalEndpoints: onlyNodeLocalEndpoints,
 	}
```
not sure what happened there but this isn't gofmt'ed correctly

timothysc commented:
re-assigning, b/c daniel is out.

dcbw commented on pkg/util/iptables/iptables.go:
```diff
@@ -575,3 +576,235 @@ func IsNotFoundError(err error) bool {
+}
+
+func parseRule(line string, filterChains []Chain, filterChainPrefixes []string) (*Rule, Chain, error) {
+	knownOptions := []knownOption{
```
Arrays can't be 'const' in Go, something about being able to take the address of an array element which would be illegal for a const.  It also shouldn't really be global since it's used in only one function, so I moved the struct declaration into the function too.

danwinship replied:
But that means it's going to be constructing the array every time you call the function. If you made it global it would only have to do it once.

dcbw commented on pkg/util/iptables/iptables.go:
```diff
@@ -575,3 +576,235 @@ func IsNotFoundError(err error) bool {
+				return nil, "", fmt.Errorf("invalid iptables rule (not enough arguments): %v", line)
+			}
+			if len(items[i]) == 0 {
+				return nil, "", fmt.Errorf("invalid iptables rule (empty argument): %v", line)
```
Fixed, multiple spaces are now handled initially by a helper that splits the line, and empty quoted arguments are passed through as-is (sans quoting of course).

dcbw commented on pkg/util/iptables/iptables.go:
```diff
@@ -575,3 +576,235 @@ func IsNotFoundError(err error) bool {
+						option.Arg += " "
+					}
+					option.Arg += items[j]
+					if strings.HasSuffix(option.Arg, "\"") {
```
Fixed.

dcbw commented on pkg/util/iptables/iptables.go:
```diff
@@ -575,3 +576,235 @@ func IsNotFoundError(err error) bool {
+		}
+
+		switch ko.option {
+		case "-A":
```
Yeah.  I just killed them, they were only used in one place in the proxy test code, and only some of them, and I don't want to add more "public" API by expanding them and moving them to iptables.go.

dcbw commented on pkg/util/iptables/iptables.go:
```diff
@@ -575,3 +576,235 @@ func IsNotFoundError(err error) bool {
+}
+
+func isChainAllowed(chain Chain, filterChains []Chain, filterChainPrefixes []string) bool {
+	if (filterChains == nil || len(filterChains) == 0) && (filterChainPrefixes == nil || len(filterChainPrefixes) == 0) {
```
Fixed.

dcbw commented on pkg/util/iptables/iptables.go:
```diff
@@ -575,3 +576,235 @@ func IsNotFoundError(err error) bool {
+				rule.Options = make(map[string]RuleOption)
+			}
+			if _, ok := rule.Options[ko.option]; ok {
+				return nil, "", fmt.Errorf("invalid iptables rule (multiple %v args): %v", ko.option, line)
```
Yeah, fixed.

dcbw commented:
@danwinship PTAL, thanks!

dcbw commented:
integration test flake is https://github.com/kubernetes/kubernetes/issues/36473

dcbw commented:
@k8s-bot test this #36473

danwinship commented on pkg/util/iptables/iptables.go:
```diff
@@ -575,3 +576,256 @@ func IsNotFoundError(err error) bool {
+}
+
+func parseRule(line string, allowChains []Chain, allowChainPrefixes []string) (*Rule, Chain, error) {
+	optionAttrs := map[string]optionAttrs{
```
as mentioned above, this has to be re-built every time you call parseRule(), while if it was global it would only have to be built once

danwinship replied:
also maybe sort them into four groups by the optionAttr values, with a comment above each group? (`// arg required, can't negate`, etc)

dcbw replied:
Done.

danwinship commented on pkg/util/iptables/iptables.go:
```diff
@@ -575,3 +576,256 @@ func IsNotFoundError(err error) bool {
+
+		attr, ok := optionAttrs[opt]
+		if !ok {
+			return nil, "", fmt.Errorf("Unknown iptables rule option %v", opt)
```
hm, I think there are some more places where you said `%v` when you meant `%q` (across multiple commits)

dcbw replied:
Done.

danwinship commented on pkg/util/iptables/iptables.go:
```diff
@@ -575,3 +576,256 @@ func IsNotFoundError(err error) bool {
+			// negation only valid at beginning of options
+			if items[i] == "!" {
+				negated = true
+				continue
```
It would simplify the loop if you just did an i++ here like you do for options with arguments. (eg, then both `opt` and `negated` would be local to the loop body)

dcbw replied:
I had it that way before, but then we have to also check if i == len(items) a lot.  I'll try it again this way and see.

danwinship commented on pkg/util/iptables/iptables_test.go:
```diff
@@ -766,3 +768,577 @@ func TestReload(t *testing.T) {
+		args, err := splitLine(tc.line)
+		if err != nil {
+			if tc.err != "" {
+				if !strings.HasPrefix(fmt.Sprintf("%v", err), tc.err) {
```
`err.Error()`. Although none of the test cases sets tc.err anyway

dcbw replied:
Done, and added some error testcases.

danwinship commented on pkg/util/iptables/iptables_test.go:
```diff
@@ -766,3 +768,577 @@ func TestReload(t *testing.T) {
+		result, err := ParseTableAddRules(tc.table, tc.filterChains, tc.filterChainPrefixes, tc.data)
+		if err != nil {
+			if tc.err != "" {
+				if !strings.HasPrefix(fmt.Sprintf("%v", err), tc.err) {
```
err.Error()

dcbw replied:
Done.

danwinship commented on pkg/proxy/iptables/proxier.go:
```diff
@@ -1292,13 +1394,24 @@ func (proxier *Proxier) syncProxyRules() {
-		// Revert new local ports.
-		revertPorts(replacementPortsMap, proxier.portsMap)
-		return
+		glog.Warningf("failed to parse iptables filter rules: %v", err)
```
With just a warning, I fear that this is just going to get broken. Someone will add a new rule that the parsing code doesn't know how to parse, no one will notice the warning message, and we'll ship a release where all of this parsing and comparing ends up just being an expensive no-op. OTOH, the code is complicated enough that I'm not comfortable saying we should panic if parsing fails...

Maybe utiliptables.EnsureRule() should return an error if it finds itself about to add a rule that it wouldn't later be able to parse? (Though in that case we'd want to validate module names too, to ensure that there aren't going to be any new "fixups for default module options" needed.)

dcbw replied:
The proxy testcases should pick up changes to the ruleset though, so it should fail pretty early if syncProxyRules() gets changed.  And it's at least designed to fail safe, as in it shouldn't change existing behavior if things don't work, it'll just always sync and log a message.

Ideally instead of adding textual rules we'd have a more structured iptables module that would use the vishvananda/netlink to read/write stuff and thus be able to more directly compare instead of screen-scraping, but that's a ton of work...

If you really want it I can try the EnsureRule thing and add hostports, userspace proxy, and other testcases too.

dcbw replied:
@danwinship the proxy code is the only one so far that uses ParseTableRules(), and while the others could use this too, they don't get called anywhere near as often so they aren't a hotpath like the proxy stuff is.  I'll definitely add comments.

dcbw commented on pkg/util/iptables/iptables.go:
```diff
@@ -575,3 +576,256 @@ func IsNotFoundError(err error) bool {
+// Split a line into space-separated arguments, handling simple quoting
+func splitLine(line string) ([]string, error) {
+	items := strings.Split(line, " ")
+	args := make([]string, 0)
```
Done.

dcbw commented:
@danwinship PTAL, thanks

dims commented:
@thockin @wojtek-t @danwinship @saad-ali : seems a bit big change at the tail end of v1.5. right?

saad-ali commented:
> @thockin @wojtek-t @danwinship @saad-ali : seems a bit big change at the tail end of v1.5. right?

Agreed, we are well into the code freeze for 1.5. Is this a bug fix? @kubernetes/sig-network for more context. Removing 1.5 milestone until we have verfication.

danwinship commented:
LGTM

@saad-ali no, this isn't a bugfix, it's a performance improvement

jeremyeder commented:
@kubernetes/sig-scalability iptables is one of Kube's lowest ceilings in terms of scalability.

bprashanth commented:
@kubernetes/sig-network is this worth reviewing for 1.5? a number of people have complained about it, so one could view it as a fix to a regression, but the regression was not introduced in 1.4

caseydavenport commented:
My 2 cents would be to hold off for v1.5 - it stinks that perf isn't great but feels risky to destabilize this late in the cycle.  I do tend to fall on the conservative side for these sorts of things, though.

Is this the sort of thing that could be a candidate for v1.5.1?

timothysc commented:
I'd consider it important enough.

thockin commented on pkg/util/iptables/iptables.go:
```diff
@@ -575,3 +576,252 @@ func IsNotFoundError(err error) bool {
+	items := strings.Split(line, " ")
+	args := make([]string, 0, len(items))
+	for i := 0; i < len(items); i++ {
+		if items[i] == "" {
```
why would this ever happen?

dcbw replied:
@thockin strings.Split("    ", " ") = [ '', '', '', '' ].  Not particularly useful, but that's a pretty common string split pattern.

thockin replied:
if you really want to split on whitespace, try `strings.Fields()`, but as described in other commentsI don't think split-on-whitespace is right.

thockin commented on pkg/util/iptables/iptables.go:
```diff
@@ -575,3 +576,252 @@ func IsNotFoundError(err error) bool {
+		items[i] = items[i][1:]
+		for j := i; j < len(items); j++ {
+			if len(arg) > 0 {
+				arg += " "
```
This is going to mutate the contents of any string that has more than one consecutive space.  Is that OK?

dcbw replied:
True, though I'd note the proxy code doesn't ever do this.  But I can certainly make that more robust.

thockin replied:
this is in `util`.  I expect others to use this, so no time bombs, please :)

I think you simply have to walk through each line, character by character.  When you find a ", you change to "read quoted string" mode until you hit that final unescaped " and change back.  Simple state-machine.

thockin replied:
That "parse a line with quoted strings" function is probably useful on its own.

thockin commented on pkg/util/iptables/iptables.go:
```diff
@@ -575,3 +576,252 @@ func IsNotFoundError(err error) bool {
+				arg += " "
+			}
+			arg += items[j]
+			if strings.HasSuffix(arg, "\"") {
```
this loop removes the outer quotes but does not unescape embedded quotes or other special characters.

e.g.  

```
# iptables -t nat -D OUTPUT -m comment --comment "foo \"bar\""

# iptables-save | grep foo
-A OUTPUT -m comment --comment "foo \"bar\""
```

you will save `foo \"bar\"`.  I would say either keep the outer quotes or run the whole thing through an unescape

thockin commented on pkg/util/iptables/iptables.go:
```diff
@@ -575,3 +576,252 @@ func IsNotFoundError(err error) bool {
+			}
+			arg += items[j]
+			if strings.HasSuffix(arg, "\"") {
+				args = append(args, arg[:len(arg)-1])
```
Actually, all my other comments be damned, this loop doesn't work, I think:

```
# iptables -t nat -I OUTPUT -m comment --comment "foo \"bar\" baz"

# iptables-save | grep foo
-A OUTPUT -m comment --comment "foo \"bar\" baz"
```

That will defeat your parser in strange ways.

thockin replied:
As above I think you simply have to walk through each line, character by character and parse by state machine

thockin commented on pkg/util/iptables/iptables.go:
```diff
@@ -575,3 +576,252 @@ func IsNotFoundError(err error) bool {
+	return nil
+}
+
+var optionAttrs = map[string]struct {
```
yeeuck, this is a race that will hurt to lose...

thockin replied:
I mean that this is an arms race with the `iptables` tool - it can add commandline opts arbitrarily, and we need to keep up.  Can we find a better pattern?

thockin replied:
Maybe we can force the caller to pass in a list of flags it uses, so the list becomes site-specific, rather than global? The list below is what kube-proxy needs, but is not generic.

thockin reviewed (commented):
I have to run now, will do more later, but I wanted to get this out ASAP.  A couple issues here.

Another issue you have to handle is that iptables will sometimes reorder arguments:

```
# iptables -t nat -I OUTPUT -m tcp -d 1.2.3.4 --dport 9376 -p tcp -j LOG

# sudo iptables-save | grep 1.2.3.4
-A OUTPUT -d 1.2.3.4/32 -p tcp -m tcp --dport 9376 -j LOG
```

Unless you sort the arguments in some canonical order (maybe you do and I missed it while skimming) this is going to be a problem.

Another problem we hit in the past is that different versions of the iptables CLI made different choices wrt this reordering and wrt quoting comments (some older versions did not quote one-word comments.

dcbw commented:
> Unless you sort the arguments in some canonical order (maybe you do and I missed it while skimming) this is going to be a problem.

@thockin the parsed arguments are stored in a map, so there is no ordering concern.  Ordering also isn't hugely important to iptables itself.  Module names are stored in a list, but that list is sorted before being compared.  So I think this is OK.

thockin commented on pkg/util/iptables/iptables.go:
```diff
@@ -575,3 +576,252 @@ func IsNotFoundError(err error) bool {
+		if !ok {
+			return nil, "", fmt.Errorf("Unknown iptables rule option %q", opt)
+		}
+		if !attr.canNegate && negated {
```
Do we actually care about this?  Why not just assume that a `!` always applies to the next arg.  If if doesn't work as an actual iptables command, who cares.  That simplifies this and more importantly the table of commands can become a map of `flagName : hasArg`

thockin commented on pkg/util/iptables/iptables.go:
```diff
@@ -575,3 +576,252 @@ func IsNotFoundError(err error) bool {
+
+			// Validate/process the required argument
+			switch opt {
+			case "-A":
```
You'd have to handle -I, -X, etc to be complete.  but...

Why bother to semantically parse?  Why not just parse it into a set of flags (and flag-arg pairs) and totally ignore any meaning?  Is that really bringing enough value to justify?

dcbw replied:
Note that iptables-save only ever outputs "-A" lines...  but we need to parse at least the top-level commands that specify which chain this line is for, since we need to be able to ignore chains that aren't owned by kube.  I'll remove the specific parsing of the others though.

thockin commented on pkg/util/iptables/iptables.go:
```diff
@@ -575,3 +576,252 @@ func IsNotFoundError(err error) bool {
+	for _, module := range rule.Modules {
+		switch module {
+		case "recent":
+			if _, ok := rule.Options["--rsource"]; !ok {
```
I don't understand these?  It does raise another issue - there are many things that the iptables command allows that get mutated on readback.  --mask becomes --xmask, and the hex value of the mask gets canonicalized.  If we can't handle that all, I am reluctant to embed this logic in a generic location - people will assume it works.

kube-proxy pre-canonicalizes mostly for historical reasons, but it would be really easy to slip up on that (or to have drift across tool versions).  Since this is distant from the use-site, it will be very hard to find.

dcbw replied:
@thockin would you prefer we pre-canonicalize everywhere?  Easy enough to do, but kinda fragile.  At least the failure case is to safe; always rewrite which is what we do now.

An alternative would be to stop constructing pure iptables-restore lines, but add actual commands to pkg/util/iptables to start a new rule and add options.  Then we can validate them and ensure they will pass our match filters. Sound OK?

thockin replied:
We can't pre-canonicalize entirely.  kube-proxy took pain to ensure it feeds args in the same order and format that relatively modern iptables reads it back, but as you saw in that other PR, older iptables is different, and I'll bet 100 bucks it changes again at some point in the future (or we get it wrong on our input).  My point is mostly - if we're going to do this, we should decide - are we generic & dumb (with opaque flags and args), or generic & smart (implementing a parser for all known iptables flags), or specialized and smart (implement a parser for all flags we know we need)?  We shouldn't be halfway.

I'm not sure I understand your alternative?

thockin reviewed (changes requested):
This is the first commit only.

thockin commented on pkg/proxy/iptables/proxier.go:
```diff
@@ -923,8 +1012,19 @@ func (proxier *Proxier) syncProxyRules() {
 	replacementPortsMap := map[localPort]closeable{}
 
+	// Sort serviceMap keys to ensure consistent order of iptables rules
+	keys := make(servicePortNameArray, 0, len(proxier.serviceMap))
```
Can we peel off little changes like this as separate PRs or commits?

thockin commented on pkg/proxy/iptables/proxier.go:
```diff
@@ -216,6 +216,15 @@ func (l *listenPortOpener) OpenLocalPort(lp *localPort) (closeable, error) {
 
+func parseMasqueradeBit(masqueradeBit int) (string, error) {
+	// Generate the masquerade mark to use for SNAT rules.
+	if masqueradeBit < 0 || masqueradeBit > 31 {
```
We should be EOL'ing this anyway, since kubelet now installs the mark chains..

thockin commented:
What conditions would cause us to have a significant number of no-op writes?

timothysc commented:
> What conditions would cause us to have a significant number of no-op writes? 

Multiple HA'd components using endpoint annotations.

dcbw commented:
> This is not right. At least on my machines, it is actually best-fit:

@thockin %0yX is supposed to zero-pad the front out to the specified number of characters.  So I don't see how "%08X" wouldn't result in 0x00000001...  Best fit is entirely contrary to the definition of "%0".  In any case with Go 1.7 on my machine, I do get zero padding.  You're saying you don't?

dcbw committed:
proxy/iptables: ensure consistent order of iptables rules

dcbw commented:
@thockin I think I've addressed all your comments except about the default module option fixups.  I'll keep thinking about that.

Turns out we don't have to hardcode options at all, since AFAIK we can just use - as a marker for option and then take whatever's next as the argument.  I've also redone the line parsing as you suggest, and removed the special handling of Comment and Jump arguments, folding them into the options dict.

dcbw commented:
@k8s-bot gci gce e2e test this issue https://github.com/kubernetes/kubernetes/issues/37361

jeremyeder commented:
To give you a sense of what we're seeing wrt iptables save/restore churn, here is a graph showing CPU utilization across a variety of system roles in a cluster.  Avg in the 2+ core range for iptables rule validation.

This cluster has 1000 nodes and 14,000 pods.  At scales far lower than this, you can find iptables consuming a core or so on every node in the cluster at any given time.  For RH'rs reading, if you'd like access please ping me internally.

![iptables](https://cloud.githubusercontent.com/assets/1919640/21182305/5e1826fc-c1d0-11e6-849a-a18131167281.png)

warden commented:
@jeremyeder out of curiosity, how many services do you have in the cluster?

jeremyeder commented:
Approximately 600 services.

widgetpl commented:
@jeremyeder do you have any network issues like packet drops etc. couse we have faced such issues when we hit to around 500 services.

jeremyeder commented:
@widgetpl Just confirmed, no drops or errors on any of the interfaces.

dcbw commented:
Going to close this one too, as the core issue may have already been addressed by other changes like https://github.com/kubernetes/kubernetes/pull/41022 and https://github.com/kubernetes/kubernetes/pull/38996
